flare/mgp/mgp.py

```diff
@@ -5,117 +5,184 @@
+                        }
     '''
 
-    def __init__(self, hyps, cutoffs, grid_params: dict, struc_params: dict,
-                 mean_only=False, container_only=True, GP=None,
-                 lmp_file_name='lmp.mgp', n_cpus=None, n_sample=100):
-
-        self.hyps = hyps
-        self.cutoffs = cutoffs
-        self.grid_params = grid_params
-        self.struc_params = struc_params
-        self.bodies = grid_params['bodies']
-        self.grid_num_2 = grid_params['grid_num_2']
-        self.bounds_2 = grid_params['bounds_2']
-        self.grid_num_3 = grid_params['grid_num_3']
-        self.bounds_3 = grid_params['bounds_3']
-
-        self.svd_rank_2 = grid_params['svd_rank_2']
-        self.svd_rank_3 = grid_params['svd_rank_3']
-        self.update = grid_params['update']
+    def __init__(self,
+                 grid_params: dict,
+                 struc_params: dict,
+                 map_force: bool=False,
+                 GP: GaussianProcess=None,
+                 mean_only: bool=False,
+                 container_only: bool=True,
+                 lmp_file_name: str='lmp.mgp',
+                 n_cpus: int=None,
+                 n_sample: int=100):
+
+        # load all arguments as attributes
+        self.map_force = map_force
         self.mean_only = mean_only
         self.lmp_file_name = lmp_file_name
         self.n_cpus = n_cpus
         self.n_sample = n_sample
+        self.grid_params = grid_params
+        self.struc_params = struc_params
+        self.hyps_mask = None
+        self.cutoffs = None
+
+        # arg_dict = inspect.getargvalues(inspect.currentframe())[3]
+        # del arg_dict['self'], arg_dict['GP']
+        # self.__dict__.update(arg_dict)
+        self.__dict__.update(grid_params)
+
+        if self.map_force and (3 in self.bodies):
+            assert (np.abs(self.bounds_3[0][2]) <= 1) or \
+                (np.abs(self.bounds_3[1][2]) <= 1), \
+                'The force mapping needs to specify [bond, bond, cos_angle] for \
+                3-body, the 3rd dimension should be in range -1 to 1'
+
+        # if GP exists, the GP setup overrides the grid_params setup
+        if GP is not None:
+
+            self.cutoffs = GP.cutoffs
+            self.hyps_mask = GP.hyps_mask
+
+            self.bodies = []
+            if "two" in GP.kernel_name:
+                self.bodies.append(2)
+                self.kernel2b_info = get_2bkernel(GP)
+            if "three" in GP.kernel_name:
+                self.bodies.append(3)
+                self.kernel3b_info = get_3bkernel(GP)
 
         self.build_bond_struc(struc_params)
         self.maps_2 = []
         self.maps_3 = []
-        self.build_map_container()
+        self.build_map_container(GP)
+        self.mean_only = mean_only
 
-        if ((not container_only) and (GP is not None) and (len(GP.training_labels_np) > 0)):
+        if not container_only and (GP is not None) and \
+                (len(GP.training_data) > 0):
```

Comment:
I see the` autorun` `bool` variable is gone. I won't die on this hill, but, one advantage of allowing the user to call the run manually means that some init variables could be come arguments passed into the `build_map` function. Right now `MappedGaussianProcess` is pretty lean, but something to keep in mind in the future.

Reply:
I delete that just because it is duplicated with `container_only`, so every time the user needs to specify `container_only` and `autorun` simultaneously. As for the variable name, I just feel that `container_only` is more explicit so I kept that one

Reply:
Ahhh yes, you're right, makes sense!